awscli/s3transfer/checksums.py

```diff
@@ -0,0 +1,222 @@
+        self._full_object_checksum = full_object_checksum
+        # If the underlying stream already has a checksum object
+        # it's updating (eg `botocore.httpchecksum.StreamingChecksumBody`),
+        # reuse its calculated value.
+        self._reuse_checksum = hasattr(self._stream, 'checksum')
+
+    @property
+    def checksum(self):
+        return self._checksum
+
+    def read(self, *args, **kwargs):
+        value = self._stream.read(*args, **kwargs)
+        if not self._reuse_checksum:
+            self._checksum.update(value)
+        if not value:
+            self._set_part_checksum()
+        return value
+
+    def _set_part_checksum(self):
+        if not self._reuse_checksum:
+            value = self._checksum.int_crc
+        else:
+            value = self._stream.checksum.int_crc
+        self._full_object_checksum.set_part_checksum(
+            self._starting_index,
+            value,
+        )
+
+
+class FullObjectChecksum:
+    def __init__(self, checksum_algorithm, content_length):
+        self.checksum_algorithm = checksum_algorithm
+        self._content_length = content_length
+        self._combine_function = _CRC_CHECKSUM_TO_COMBINE_FUNCTION[
+            self.checksum_algorithm
+        ]
+        self._stored_checksum = None
+        self._part_checksums = None
+        self._calculated_checksum = None
+
+    @cached_property
+    def calculated_checksum(self):
+        if self._calculated_checksum is None:
+            self._combine_part_checksums()
+        return self._calculated_checksum
+
+    def set_stored_checksum(self, stored_checksum):
+        self._stored_checksum = stored_checksum
+
+    def set_part_checksum(self, offset, checksum):
+        if self._part_checksums is None:
+            self._part_checksums = {}
+        self._part_checksums[offset] = checksum
+
+    def _combine_part_checksums(self):
+        if self._part_checksums is None:
+            return
+
+        sorted_offsets = sorted(self._part_checksums.keys())
+        # Initialize the combined checksum to the first part's checksum value.
+        combined = self._part_checksums[sorted_offsets[0]]
+        # To calculate part length, take the current offset and subtract from
+        # the next offset. If the current offset is the start of the last part,
+        # then subtract from the total content length. eg,
+        # (8388608, 16777216)
+        # (16777216, self._content_length)
+        remaining_offsets = sorted_offsets[1:]
+        next_offsets = sorted_offsets[2:] + [self._content_length]
+        for offset, next_offset in zip(remaining_offsets, next_offsets):
+            part_checksum = self._part_checksums[offset]
+            offset_len = next_offset - offset
+            combined = self._combine_function(
+                combined, part_checksum, offset_len
+            )
+
+        self._calculated_checksum = base64.b64encode(
+            combined.to_bytes(4, byteorder='big')
+        ).decode('ascii')
```

Comment:
This is actually dependent on the checksum algorithm. I think instead I'll instantiate the appropriate checksum object, set the `int_crc` property to the combined value, then call the `digest()` method